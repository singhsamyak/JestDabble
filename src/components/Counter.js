import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Counter extends Component {
  render() {
    const { value, increment, decrement, reset } = this.props;
    return(
      <div>
        <p>Count: <span>{ value }</span> times</p>
        <button id='increment' onClick={ increment }>+</button>
        <button id='decrement' onClick={ decrement }>-</button>
        <button id='reset' onClick={ reset }>Reset</button>
      </div>
    );
  }
}

Counter.PropTypes = {
  value: PropTypes.number.isRequired,
  increment: PropTypes.func.isRequired,
  decrement: PropTypes.func.isRequired,
  reset: PropTypes.func.isRequired
}

export default Counter;
