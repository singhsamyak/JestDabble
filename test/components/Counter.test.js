import React from 'react';
import Counter from '../../src/components/Counter';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

/**
 * Mock Functions
 * ---------------
 * - Erases the actual implementation of a function
 * - Captures calls to the function
 * - Captures instances of constructor functions when instantiated with `new`
 * - Allows test-time configuration of return values
 */
describe('Snapshot Testing using Jest only', () => {
  it ('renders correctly', () => {
    const value = 0;
    const actions = {
      increment: jest.fn(), // jest.fn is a mock function
      decrement: jest.fn(),
      reset    : jest.fn()
    }
    const tree = renderer.create(
      <Counter value={value} {...actions} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

function setup(value = 0) {
  const actions = {
    increment: jest.fn(), // jest.fn is a mock function
    decrement: jest.fn(),
    reset    : jest.fn()
  }
  const counter = shallow(
    <Counter value={value} {...actions} />
  );

  return {
    counter  : counter,
    actions  : actions,
    incrBtn  : counter.find('#increment'),
    decrBtn  : counter.find('#decrement'),
    resetBtn : counter.find('#reset'),
    countVal : counter.find('span')
  };
}

/**
 * Enzyme's Shallow Rendering
 * --------------------------
 * - Helpful to test a component as a unit.
 * - Ensures test aren't asserting based on behavior of child components
 * - http://airbnb.io/enzyme/docs/api/shallow.html
 */
describe('DOM Testing using Enzyme', () => {
  it ('should initially have a count value of 0', () => {
    const { countVal } = setup();
    expect(countVal.text()).toEqual('0');
  });

  it ('clicking + should invoke the increment callback', () => {
    const { actions, incrBtn } = setup();

    incrBtn.simulate('click');
    expect(actions.increment).toBeCalled();
    expect(actions.decrement).not.toBeCalled();
  });

  it ('clicking - should invoke the decrement callback', () => {
    const { actions, decrBtn } = setup();

    decrBtn.simulate('click');
    expect(actions.decrement).toBeCalled();
    expect(actions.increment).not.toBeCalled();
  });

  it ('clicking Reset should invoke the reset callback', () => {
    const { actions, resetBtn } = setup();

    resetBtn.simulate('click');
    expect(actions.reset).toBeCalled();
    expect(actions.decrement).not.toBeCalled();
    expect(actions.increment).not.toBeCalled();
  });
});
