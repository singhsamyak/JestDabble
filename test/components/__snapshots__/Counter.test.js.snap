// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing using Jest only renders correctly 1`] = `
<div>
  <p>
    Clicked: 
    <span>
      0
    </span>
     times
  </p>
  <button
    id="increment"
    onClick={[Function]}
  >
    +
  </button>
  <button
    id="decrement"
    onClick={[Function]}
  >
    -
  </button>
  <button
    id="reset"
    onClick={[Function]}
  >
    Reset
  </button>
</div>
`;
